ntoskrnl/io/iomgr/deviface.c

```diff
@@ -30,185 +31,421 @@ static PWCHAR BaseKeyString = L"\\Registry\\Machine\\System\\CurrentControlSet\\
+    DeviceString.MaximumLength = LinkNameNoPrefix.Length - ReferenceStringReal.Length;
+    DeviceString.Length = DeviceString.MaximumLength;
+    DeviceString.Buffer = LinkNameNoPrefix.Buffer;
+
+    DPRINT("Device string is '%wZ'\n", &DeviceString);
+
+    /* Allocate a munged path string */
+    MungedStringReal.Length = 0;
+    MungedStringReal.MaximumLength = MungedPrefix.Length + DeviceString.Length + sizeof(UNICODE_NULL);
+    MungedStringReal.Buffer = ExAllocatePoolWithTag(NonPagedPool, MungedStringReal.MaximumLength, TAG_IO);
+    if (!MungedStringReal.Buffer)
+    {
+        DPRINT1("ExAllocatePoolWithTag() failed\n");
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    /* Fill it by munged prefix and munged string */
+    Status = RtlAppendUnicodeStringToString(&MungedStringReal, &MungedPrefix);
+    NT_ASSERT(NT_SUCCESS(Status));
+
+    Status = RtlAppendUnicodeStringToString(&MungedStringReal, &DeviceString);
+    NT_ASSERT(NT_SUCCESS(Status));
+
+    DPRINT("Munged string is '%wZ'\n", &MungedStringReal);
+    DPRINT("GUID string is '%wZ'\n", &GuidStringReal);
+    DPRINT("Reference string is '%wZ'\n", &ReferenceStringReal);
+
+    /* Store received parts if the parameters are not null */
+    if (PrefixString != NULL)
+    {
+        if (RtlPrefixUnicodeString(&KernelModePrefix, SymbolicLinkName, FALSE))
+            *PrefixString = KernelModePrefix;
+        else if (RtlPrefixUnicodeString(&UserModePrefix, SymbolicLinkName, FALSE))
+            *PrefixString = UserModePrefix;
+    }
+
+    if (MungedString != NULL)
+        *MungedString = MungedStringReal;
     else
-        InstanceKeyRealP = &InstanceKeyReal;
+        RtlFreeUnicodeString(&MungedStringReal);
 
-    *GuidKeyRealP = NULL;
-    *DeviceKeyRealP = NULL;
-    *InstanceKeyRealP = NULL;
+    if (GuidString != NULL)
+        *GuidString = GuidStringReal;
 
-    RtlInitUnicodeString(&BaseKeyU, BaseKeyString);
+    if (ReferenceString != NULL)
+        *ReferenceString = ReferenceStringReal;
 
-    /* Open the DeviceClasses key */
-    InitializeObjectAttributes(&ObjectAttributes,
-                               &BaseKeyU,
-                               OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
-                               NULL,
-                               NULL);
-    Status = ZwOpenKey(&ClassesKey,
-                       DesiredAccess | KEY_ENUMERATE_SUB_KEYS,
-                       &ObjectAttributes);
+    if (ReferenceStringPresent != NULL)
+        *ReferenceStringPresent = ReferenceStringReal.Length != sizeof(WCHAR);
+
+    if (InterfaceClassGuid != NULL)
+    {
+        /* Convert GUID string into a GUID and store it also */
+        Status = RtlGUIDFromString(&GuidStringReal, InterfaceClassGuid);
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("RtlGUIDFromString() failed with status 0x%08lx\n", Status);
+            RtlFreeUnicodeString(&MungedStringReal);
```

Comment:
Since we've already set the output parameter at this point, it would be best practice to re-set it to null rather than leaving it pointing to something invalid, i.e. `RtlInitEmptyUnicodeString(MungedString, NULL, 0)`.

Reply:
Fixed.